### ISSUE TITLE
Recurring Subscription Payments failing with errors for some customers

### ISSUE THREAD

**Issue by stuartduff** [Automattic/woocommerce-payments#4829]
### Describe the bug

It seems this commit below within v4.8.0 of WooCommerce Payments is causing subscription renewal payments and retries to fail for customers using WooCommerce Payments included Subscriptions features.

**Commit:** https://github.com/Automattic/woocommerce-payments/commit/e25a7e55cdcae2d407fd2a59753e6526e8ecfe54

### To Reproduce

I have not as yet been able to reproduce the error but we've now seen multiple reports of the same error message in support. I've referenced those tickets within this bug report.

### Actual behaviour

Subscription renewals are failing and producing an error message like the one below within s sites' failed scheduled action logs. Also on the scheduled actions failed screen related to that action.

```
10-01-2022 @ 09:52:38 - scheduled action 75569 (subscription payment) failed to finish processing due to the following error: Uncaught Error: Class "WC_Payments_Subscription_Service" not found in /srv/htdocs/wp-content/plugins/woocommerce-payments/includes/compat/subscriptions/trait-wc-payment-gateway-wcpay-subscriptions.php:270
Stack trace:
#0 /wordpress/core/6.0.2/wp-includes/class-wp-hook.php(307): WC_Payment_Gateway_WCPay->scheduled_subscription_payment('45.00', Object(Automattic\WooCommerce\Admin\Overrides\Order))
#1 /wordpress/core/6.0.2/wp-includes/class-wp-hook.php(331): WP_Hook->apply_filters('', Array)
#2 /wordpress/core/6.0.2/wp-includes/plugin.php(476): WP_Hook->do_action(Array)
#3 /srv/htdocs/wp-content/plugins/woocommerce-subscriptions/includes/gateways/class-wc-subscriptions-payment-gateways.php(97): do_action('woocommerce_sch...', '45.00', Object(Automattic\WooCommerce\Admin\Overrides\Order))
#4 /srv/htdocs/wp-content/plugins/woocommerce-subscriptions/includes/gateways/class-wc-subscriptions-payment-gateways.php(81): WC_Subscriptions_Payment_Gateways::trigger_gateway_renewal_payment_hook(Object(Automattic\WooCommerce\Admin\Overrides\Order))
#5 /wordpress/core/6.0.2/wp-includes/class-wp-hook.php(307): WC_Subscriptions_Payment_Gateways::gateway_scheduled_subscription_payment(70991)
#6 /wordpress/core/6.0.2/wp-includes/class-wp-hook.php(331): WP_Hook->apply_filters(NULL, Array)
#7 /wordpress/core/6.0.2/wp-includes/plugin.php(524): WP_Hook->do_action(Array)
#8 /wordpress/plugins/woocommerce/6.9.4/packages/action-scheduler/classes/actions/ActionScheduler_Action.php(22): do_action_ref_array('woocommerce_sch...', Array)
#9 /wordpress/plugins/woocommerce/6.9.4/packages/action-scheduler/classes/abstracts/ActionScheduler_Abstract_QueueRunner.php(65): ActionScheduler_Action->execute()
#10 /wordpress/plugins/woocommerce/6.9.4/packages/action-scheduler/classes/ActionScheduler_QueueRunner.php(162): ActionScheduler_Abstract_QueueRunner->process_action(75569, 'WP Cron')
#11 /wordpress/plugins/woocommerce/6.9.4/packages/action-scheduler/classes/ActionScheduler_QueueRunner.php(132): ActionScheduler_QueueRunner->do_batch(25, 'WP Cron')
#12 /wordpress/core/6.0.2/wp-includes/class-wp-hook.php(307): ActionScheduler_QueueRunner->run('WP Cron')
#13 /wordpress/core/6.0.2/wp-includes/class-wp-hook.php(331): WP_Hook->apply_filters('', Array)
#14 /wordpress/core/6.0.2/wp-includes/plugin.php(524): WP_Hook->do_action(Array)
#15 /wordpress/core/6.0.2/wp-cron.php(138): do_action_ref_array('action_schedule...', Array)
#16 {main}
  thrown
10-01-2022 @ 09:52:38 - action args: subscription_id: 70991
```

### Screenshots

<img width="398" alt="Screen Shot 2022-10-01 at 15 21 25" src="https://user-images.githubusercontent.com/1190565/193414021-15e3d2b5-eb35-4842-9913-094b63c1a829.png">

### Expected behavior

Subscriptions should renew without any such error.

### Additional context

(Edit by @csmcneill) Affected users so far:

- 5598535-zen
- 5598822-zen
- 5598193-zen
- 5597793-zen
- 5597095-zen
- 5598038-zen
- 5598172-zen
- 5599508-zen
- 5599021-zen
- 5599314-zen
- 5598920-zen
- 5599166-zen
- 37209933-hc
- 5601614-zen
- 5599941-zen
- 33607658-hc | 5601965-zen
- 36758763-hc
- 5602994-zen
- 5597209-zen
- 5604379-zen
- 5604140-zen
- 37236321-hc
- 5604768-zen
- 5602954-zen

**Comment by fernmed** [comment]
5598038-zen 
5598172-zen

having same error appear on multiple subscription renewals.

**Comment by jnineb** [comment]
5599508-zen 

Having the same error occur for multiple subscriptions using WCPay

**Comment by csmcneill** [comment]
@stuartduff @fernmed @jnineb do these affected users have WC Subs and are using WCPay as the gateway? Or are they just using WCPay Subs?

**Comment by csmcneill** [comment]
cc’ing @james-allan here since this appears to be related to his commit.

**Comment by jnineb** [comment]
reported in [5599021-zen](5599021-zd-woothemes)
Using WC Subscriptions and WCPay as the payment gateway

**Comment by fernmed** [comment]
@csmcneill Of the ones I found, and the ones I observed in the queue that Stuart assisted on, they all are using WC Subs and WC Pay as the gateway.

**Comment by jnineb** [comment]
> do these affected users have WC Subs and are using WCPay as the gateway? Or are they just using WCPay Subs?

@csmcneill The two I've come across were both using WooCommerce Subscriptions and WCPay as the payment gateway. [5599508-zen](5599508-zd-woothemes) has subscriptions also processed via Stripe and PayPal, which have been unaffected.

**Comment by jnineb** [comment]
reported again in 5599314-zen they indicated the issue began after recently updating WCPay

**Comment by AbhinavSakalle** [comment]
Reported in 5598920-zen. Retrying a manual renewal didn't work for them.

**Comment by ChrissiePollock** [comment]
#5599166-zen said " It looked like it wanted me to activate both Woo Payments AND Stripe on the plugins screen. I just turned off Stripe and kept Woo Payments activated. Still the same problems, though." Their Stripe had been disconnected earlier (not by them -- that was a glitch as well. Not sure if that is related to this issue.)

**Comment by csmcneill** [comment]
37209933-hc

```
2022-10-02T11:44:48+00:00 CRITICAL Uncaught Error: Class 'WC_Payments_Subscription_Service' not found in /home/yg4c5xy7yym2/public_html/wp-content/plugins/woocommerce-payments/includes/compat/subscriptions/trait-wc-payment-gateway-wcpay-subscriptions.php:270
Stack trace:
#0 /home/yg4c5xy7yym2/public_html/wp-includes/class-wp-hook.php(307): WC_Payment_Gateway_WCPay->scheduled_subscription_payment('25.00', Object(Automattic\WooCommerce\Admin\Overrides\Order))
#1 /home/yg4c5xy7yym2/public_html/wp-includes/class-wp-hook.php(331): WP_Hook->apply_filters('', Array)
#2 /home/yg4c5xy7yym2/public_html/wp-includes/plugin.php(476): WP_Hook->do_action(Array)
#3 /home/yg4c5xy7yym2/public_html/wp-content/plugins/woocommerce-subscriptions/includes/gateways/class-wc-subscriptions-payment-gateways.php(97): do_action('woocommerce_sch...', '25.00', Object(Automattic\WooCommerce\Admin\Overrides\Order))
#4 /home/yg4c5xy7yym2/public_html/wp-content/plugins/woocommerce-subscriptions/includes/gateways/class-wc-subscriptions-payment-gateways.php(81): WC_Subscriptions_ in /home/yg4c5xy7yym2/public_html/wp-content/plugins/woocommerce-payments/includes/compat/subscriptions/trait-wc-payment-gateway-wcpay-subscriptions.php on line 270
```

Edit: Reverting to 4.7.2 and retrying the renewal payment worked:

<img width="321" alt="Markup 2022-10-02 at 11 27 29" src="https://user-images.githubusercontent.com/82999806/193468190-479d86b6-bd35-4eb8-b0b0-4ab939534286.png">

**Comment by haszari** [comment]
Linking in more context. The linked commit is from PR #4790. This is a fix to prevent WCPay Subscriptions (stripe-billing subscriptions) double-charging when the site has WC Subscriptions active. This scenario would apply to sites that switched to WC Subscriptions (e.g. upgraded from built-in free option to paid option). 

- #4788 

Couple of possible issues here to investigate:

- Dependency issue causing fatal error `'WC_Payments_Subscription_Service' not found`. This seems most likely.
- Logic error in the changed code preventing renewal processing. Less likely but want to confirm this is is not an issue.

Next step:

- **Confirm reproduce steps.** As I understand it we have not yet reproduced the issue, though if it's a fatal / dependency issue might be straightforward.

**Comment by EjayhanFernandes** [comment]
> 37209933-hc
> 
> ```
> 2022-10-02T11:44:48+00:00 CRITICAL Uncaught Error: Class 'WC_Payments_Subscription_Service' not found in /home/yg4c5xy7yym2/public_html/wp-content/plugins/woocommerce-payments/includes/compat/subscriptions/trait-wc-payment-gateway-wcpay-subscriptions.php:270
> Stack trace:
> #0 /home/yg4c5xy7yym2/public_html/wp-includes/class-wp-hook.php(307): WC_Payment_Gateway_WCPay->scheduled_subscription_payment('25.00', Object(Automattic\WooCommerce\Admin\Overrides\Order))
> #1 /home/yg4c5xy7yym2/public_html/wp-includes/class-wp-hook.php(331): WP_Hook->apply_filters('', Array)
> #2 /home/yg4c5xy7yym2/public_html/wp-includes/plugin.php(476): WP_Hook->do_action(Array)
> #3 /home/yg4c5xy7yym2/public_html/wp-content/plugins/woocommerce-subscriptions/includes/gateways/class-wc-subscriptions-payment-gateways.php(97): do_action('woocommerce_sch...', '25.00', Object(Automattic\WooCommerce\Admin\Overrides\Order))
> #4 /home/yg4c5xy7yym2/public_html/wp-content/plugins/woocommerce-subscriptions/includes/gateways/class-wc-subscriptions-payment-gateways.php(81): WC_Subscriptions_ in /home/yg4c5xy7yym2/public_html/wp-content/plugins/woocommerce-payments/includes/compat/subscriptions/trait-wc-payment-gateway-wcpay-subscriptions.php on line 270
> ```
> 
> Edit: Reverting to 4.7.2 and retrying the renewal payment worked:
> 
> <img alt="Markup 2022-10-02 at 11 27 29" width="321" src="https://user-images.githubusercontent.com/82999806/193468190-479d86b6-bd35-4eb8-b0b0-4ab939534286.png">

Confirming that this also worked in 5597793-zen.

**Comment by EjayhanFernandes** [comment]
Another report here: 5601614-zen. Asking them to roll back to version `4.7.2`.

**Comment by thisissandip** [comment]
5599941-zen. Asked them to roll back to version 4.7.2 and retry the renewal payment

**Comment by thisissandip** [comment]
Another report: 33607658-hc | 5601965-zen

**Comment by RadoslavGeorgiev** [comment]
Moving this issue to the _Other Teams_ queue in the WCPay backlog, because it is being looked into: p1664761278711219-slack-C02BW3Z8SHK

@haszari there seem to be a lot of reports, ping me if you need help triaging this.

**Comment by csmcneill** [comment]
36758763-hc

**Comment by kingsleyinfo** [comment]
Also seen here: 5602994-zen

Suggested the workaround

**Comment by csmcneill** [comment]
5597209-zen

**Comment by johndcoy** [comment]
Another in 5603014-zen 

This ticket has PHP and DB logs

/var/log/litespeed/php_errors.log
/var/log/mysql/mysqld.log

**Comment by haszari** [comment]
@stuartduff  @kingsleyinfo @csmcneill @thisissandip Has anyone confirmed reproduce steps for this issue yet? I tried on my test site yesterday and couldn't reproduce the issue. Here's the flow I tested:

- _Set up a subscription…_
- Activate WC Subscriptions and WCPay (4.8) as payment gateway – using subscriptions integration
- Make a subscription product (1 day renewal)
- Purchase a subscription
- _Force a renewal using admin…_
- Go to `WooCommerce > Subscriptions`
    - Click on subscription id
    - Select `Create pending renewal order` from actions and click `Update`
- Go to new `WooCommerce > Orders`
    - Click on renewal order
    - Select `Retry renewal payment` from actions and click `Update`

Since this is a fatal error it should be relatively quick to fix, as long as we can reproduce the issue. @brucealdridge  can you investigate the tickets and figure out how to reproduce?

**Comment by AbhinavSakalle** [comment]
@haszari attempting a renewal resulted in "Fatal error: Uncaught Error: Class 'WC_Payments_Subscription_Service' not found"
Path: wp-content/plugins/woocommerce-payments/includes/compat/subscriptions/trait-wc-payment-gateway-wcpay-subscriptions.php on line 270

**Comment by AbhinavSakalle** [comment]
5604379-zen

**Comment by csmcneill** [comment]
Our engineering team has determined that if `WC_Payments_Features::is_wcpay_subscriptions_enabled` returns `false`, and a subscription renews, this triggers the error.

`WC_Payments_Features::is_wcpay_subscriptions_enabled` returns `false` if `Enable Subscriptions with WooCommerce Payments` is not checked in the WCPay advanced settings. This option is only available for US accounts, so it will always return `false` for non-US accounts.

I was able to reproduce this on my own US-based test site with the following steps:

1. Disable WC Subscription (the plugin).
2. Go to the WC Payments advanced settings
3. Uncheck the “Enable Subscriptions with WooCommerce Payments” setting
4. Enable WC Subscriptions
5. Purchase a subscription
6. Go to the edit subscription screen.
7. Process a renewal via the actions dropdown.

#4839 is going through testing right now, and it looks like a candidate for a patch. Stay tuned.

**Comment by nicdwilson** [comment]
5604140-zen

**Comment by francoishvz** [comment]
37236321-hc Asked them to roll back to version 4.7.2 and retry the renewal payment which worked.

**Comment by stuartduff** [comment]
5604768-zen Provided steps on how to roll back to v4.7.2

**Comment by imodouglas** [comment]
5602954-zen is affected as well. Found [this Droplr link](https://d.pr/f/hennYe) with the v4.7.2 file. Hope this helps!

**Comment by csmcneill** [comment]
@imodouglas Since WCPay is a public repo, all releases are available directly from GitHub. They're also available from WordPress.org, which means there's no real need to rely on an external file-sharing service to share older versions of the software.

- GitHub link to 4.7.2: https://github.com/Automattic/woocommerce-payments/releases/download/4.7.2/woocommerce-payments.zip
- WordPress.com link to 4.7.2: https://downloads.wordpress.org/plugin/woocommerce-payments.4.7.2.zip

**Comment by cesarcosta99** [comment]
Closing this issue since the fix is merged and released in [4.8.1](https://wordpress.org/plugins/woocommerce-payments/).
